### DISCUSSION TITLE
Q: Example of how to execute IRepositoryClient.GetArchive()?

### DISCUSSION THREAD

**Discussion by UsgsMonkey** [ubisoft/NGitLab#677] [Q&A]
Greetings,

I'm relatively new to asynchronous programming and am looking for guidance or links to proper guidance on how to code in the "Action\<Stream\>" argument input required by IRepositoryClient.GetArchive().  This seems to reference an async pattern and there's not documentation to lead the way.  Any assistance would be a huge help as I stumble about trying to figure out the expected approach here.  Ultimately, I'm trying to use NGitLab as a vehicle to download an archive (as a zip or other compressed format) to disk programmatically.

Warm Regards,

Brett

**Comment by louis-z** [comment]
Hi @UsgsMonkey,

Although NGitLab does offer async methods, `IRepositoryClient.GetArchive(Action<Stream> parser)` is not async. The `parser` argument is used to pass in a callback to handle the `tar.gz` stream returned by GitLab. For instance, based on the code snippet found [here](https://www.nikouusitalo.com/blog/how-to-natively-read-tgz-files-with-the-new-c-tarreader-class/) you could print out the content of the archive as follows:

```cs
RepositoryClient.GetArchive(tgzStream =>
{
    using var gzip = new GZipStream(tgzStream, CompressionMode.Decompress);

    using var unzippedStream = new MemoryStream();
    gzip.CopyTo(unzippedStream);
    unzippedStream.Seek(0, SeekOrigin.Begin);

    using var reader = new TarReader(unzippedStream);

    while (reader.GetNextEntry() is TarEntry entry)
    {
        Console.WriteLine($"Entry name: {entry.Name}, entry type: {entry.EntryType}");
    }
});
```
